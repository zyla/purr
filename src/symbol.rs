#[derive(PartialEq, Eq, PartialOrd, Ord, Debug, Clone, Hash)]
pub struct Symbol(pub String);

impl Symbol {
    pub fn new(s: String) -> Self {
        Self(s)
    }
}
