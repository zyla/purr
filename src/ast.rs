use crate::symbol::Symbol;

#[derive(Debug)]
pub struct SourceSpan {
    pub start: usize,
    pub end: usize,
}

#[derive(Debug)]
pub struct Located<T>(pub SourceSpan, pub T);

#[derive(Debug)]
pub struct Comment(pub String);

#[derive(Debug)]
pub struct Commented<T>(pub Vec<Comment>, pub T);

pub type Module = Located<Commented<ModuleInner>>;

#[derive(Debug)]
pub struct ModuleInner {
    pub name: Symbol,
    pub exports: Option<Vec<DeclarationRef>>,
    pub declarations: Vec<Declaration>,
}

#[derive(Debug)]
pub enum DeclarationRef {
    // TODO
}

pub type Declaration = Located<Commented<DeclarationKind>>;

#[derive(Debug)]
pub enum DeclarationKind {
    ValueDeclaration(ValueDeclaration),
}

#[derive(Debug)]
pub struct ValueDeclaration {
    pub ident: Symbol,

    // TODO: what is this?
    // pub name: NameKind

    // TODO
    // pub binders: Vec<Binder>
    pub expr: Vec<GuardedExpr>,
}

#[derive(Debug)]
pub struct GuardedExpr {
    pub guards: Vec<Guard>,
    pub expr: Expr,
}

#[derive(Debug)]
pub enum Guard {
    // TODO
}

pub type Expr = Located<ExprKind>;

#[derive(Debug)]
pub enum ExprKind {
    Literal(Literal<Expr>),
    // TODO
}

#[derive(Debug)]
pub enum Literal<T> {
    Integer(i32),
    Float(f64),
    String(String),
    Char(char),
    Boolean(bool),
    Array(Vec<T>),
    Object(Vec<(Symbol, T)>),
}
