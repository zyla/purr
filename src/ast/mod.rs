pub mod declarations;
pub mod expr;
pub mod meta;
pub mod types;

use crate::ModuleId;
pub use declarations::*;
pub use expr::*;
pub use meta::*;
pub use types::*;

use crate::symbol::Symbol;

#[salsa::interned]
pub struct AbsoluteName {
    pub module: ModuleId,
    pub name: Symbol,
}

#[salsa::interned]
pub struct QualifiedName {
    pub module: Option<ModuleId>,
    pub name: Symbol,
}

impl QualifiedName {
    pub fn is_actually_qualified(&self, db: &dyn crate::Db) -> bool {
        self.module(db).is_some()
    }

    pub fn new_unqualified(db: &dyn crate::Db, name: Symbol) -> Self {
        Self::new(db, None, name)
    }

    pub fn new_qualified(db: &dyn crate::Db, module: ModuleId, name: Symbol) -> Self {
        Self::new(db, Some(module), name)
    }
}
