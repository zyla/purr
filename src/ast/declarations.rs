use super::{Commented, Located, Pat, PossiblyGuardedExpr, Type};
use crate::ast::QualifiedName;
use crate::symbol::Symbol;

pub type Module = Located<Commented<ModuleInner>>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct ModuleInner {
    pub name: ModuleName,
    pub exports: Option<Vec<DeclarationRef>>,
    pub imports: Vec<Import>,
    pub declarations: Vec<Declaration>,
}

pub type Import = Located<ImportInner>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct ImportInner {
    pub module: ModuleName,
    pub kind: ImportDeclarationKind,
    pub alias: Option<ModuleName>,
}

pub type DeclarationRef = Located<DeclarationRefKind>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum NameSource {
    UserNamed,
    CompilerNamed,
}

pub type ModuleName = QualifiedName;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum DeclarationRefKind {
    TypeClass {
        name: Symbol,
    },
    TypeOp {
        name: Symbol,
    },
    Type {
        name: Symbol,
        constructors: Option<DeclarationRefConstructors>,
    },
    Value {
        name: Symbol,
    },
    ValueOp {
        name: Symbol,
    },
    TypeInstanceRef {
        name: Symbol,
        name_source: NameSource,
    },
    Module {
        name: ModuleName,
    },
    ReExport {
        imported_from: ModuleName,
        defined_in: ModuleName,
        declaration_ref: Box<DeclarationRef>,
    },
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum DeclarationRefConstructors {
    All,
    Some(Vec<Symbol>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum ImportDeclarationKind {
    Implicit,
    Explicit(Vec<DeclarationRef>),
    Hiding(Vec<DeclarationRef>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct RoleDeclarationData {
    pub ident: Symbol,
    pub role: Vec<Role>,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum Role {
    Nominal,
    Representational,
    Phantom,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct TypeDeclarationData {
    pub ident: Symbol,
    pub r#type: Type,
}

impl TypeDeclarationData {
    pub fn new(ident: Symbol, r#type: Type) -> Self {
        Self { ident, r#type }
    }
}

pub type Declaration = Located<Commented<DeclarationKind>>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum DeclarationKind {
    Data {
        type_: DataDeclType,
        name: Symbol,
        params: Vec<TypeParameter>,
        kind: Option<Type>,
        constructors: Vec<DataConstructorDeclaration>,
    },

    // TODO: do we need this? seems internal
    DataBindingGroup(Vec<Declaration>),

    TypeSynonym {
        name: Symbol,
        params: Vec<TypeParameter>,
        body: Type,
    },

    KindSignature {
        for_type: KindSignatureFor,
        name: Symbol,
        kind: Kind,
    },

    Role(RoleDeclarationData),

    TypeSignature(TypeDeclarationData),

    ValueDeclaration(ValueDeclaration),

    // C x y = z, used in `let`
    Destructuring {
        pat: Pat,
        expr: PossiblyGuardedExpr,
    },

    ForeignValue {
        name: Symbol,
        type_: Type,
    },

    Class(TypeClassDeclaration),

    InstanceChain(Vec<InstanceDeclaration>),

    Operator {
        associativity: Associativity,
        precedence: u8,
        name: OperatorTarget,
        operator: Symbol,
    },
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum Associativity {
    None,
    Left,
    Right,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum OperatorTarget {
    Type(QualifiedName),
    DataConstructor(QualifiedName),
    Value(QualifiedName),
}

pub type Kind = Type;

pub type TypeParameter = (Symbol, Option<Kind>);

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct TypeClassDeclaration {
    pub constraints: Vec<Type>,
    pub name: Symbol,
    pub params: Vec<TypeParameter>,
    pub fundeps: Vec<Fundep>,
    pub methods: Vec<TypeDeclarationData>,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct InstanceDeclaration {
    pub constraints: Vec<Type>,
    pub instance_type: InstanceType,
    pub instance_name: Option<Symbol>,
    pub class: QualifiedName,
    pub args: Vec<Type>,
    pub body: Vec<Declaration>,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct Fundep {
    pub from: Vec<Symbol>,
    pub to: Vec<Symbol>,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum InstanceType {
    Plain,
    Derive,
    DeriveNewtype,
}

// Note: `data` and `newtype` signatures are actually declarations without constructors
#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum KindSignatureFor {
    TypeSynonym,
    Class,
}

pub type DataConstructorDeclaration = Located<Commented<DataConstructorDeclarationData>>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct DataConstructorDeclarationData {
    pub name: Symbol,

    // TODO: in original AST they have names. Why? I thought datacon fields are unnamed in PS
    pub fields: Vec<Type>,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum DataDeclType {
    Data,
    ForeignData,
    Newtype,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct ValueDeclaration {
    pub ident: Symbol,

    // TODO: what is this?
    // pub name: NameKind
    pub params: Vec<Pat>,
    pub expr: PossiblyGuardedExpr,
}
