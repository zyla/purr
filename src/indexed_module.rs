use crate::ast::AbsoluteName;
use crate::ast::CaseBranch;
use crate::ast::DataConstructorDeclaration;
use crate::ast::DataDeclType;
use crate::ast::Declaration;
use crate::ast::DeclarationKind;
use crate::ast::Fundep;
use crate::ast::Located;
use salsa::DebugWithDb;
use std::iter::Peekable;
use std::path::PathBuf;

use crate::ast::Type;
use crate::ast::TypeDeclarationData;
use crate::ast::TypeParameter;
use crate::ParsedModule;
use crate::{Db, ModuleId};
use crate::{Diagnostic, Diagnostics};
use fxhash::FxHashMap;
use std::collections::hash_map::Entry;

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb)]
pub enum TypeDecl {
    Data(DataDecl),
    Type(TypeSynonymDecl),
    TypeClass(TypeClassDecl),
}

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb)]
pub struct DataDecl {
    pub type_: DataDeclType,
    pub name: AbsoluteName,
    pub params: Vec<TypeParameter>,
    pub kind: Option<Type>,
    pub constructors: Vec<DataConstructorDeclaration>,
}

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb)]
pub struct TypeSynonymDecl {
    pub name: AbsoluteName,
    pub params: Vec<TypeParameter>,
    pub body: Type,
}

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb)]
pub struct TypeClassDecl {
    pub name: AbsoluteName,
    pub constraints: Vec<Type>,
    pub params: Vec<TypeParameter>,
    pub fundeps: Vec<Fundep>,
    pub methods: Vec<TypeDeclarationData>,
}

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb)]
pub struct ValueDecl {
    pub name: AbsoluteName,
    pub type_: Option<Type>,

    /// Empty equations means foreign import.
    /// TODO: maybe make it more explicit?
    pub equations: Vec<CaseBranch>,
}

struct ModuleIndexer<'a> {
    db: &'a dyn Db,
    // Note: Using `FxHashMap` mostly because we want deterministic order for snapshots.
    types: FxHashMap<AbsoluteName, TypeDecl>,
    values: FxHashMap<AbsoluteName, ValueDecl>,
    classes: FxHashMap<AbsoluteName, TypeClassDecl>,
    module_id: ModuleId,
    filename: PathBuf,
    // TODO: Add map with Decl to SourceSpan
}

impl<'a> ModuleIndexer<'a> {
    fn index(&mut self, module: ParsedModule) {
        let db = self.db;
        let mut iter = module.ast.declarations.iter().peekable();
        while let Some(src_decl) = iter.peek().copied() {
            use DeclarationKind::*;

            let ref_loc = src_decl.0.start;

            match &***src_decl {
                Data {
                    type_,
                    name,
                    params,
                    kind,
                    constructors,
                } => {
                    let abs_name = AbsoluteName::new(db, self.module_id, *name);
                    let mut params = params.clone();
                    to_relative_params(abs_name, ref_loc, &mut params);

                    let mut constructors = constructors.clone();
                    constructors
                        .iter_mut()
                        .for_each(|c| c.to_relative_span(abs_name, ref_loc));

                    match self.types.entry(abs_name) {
                        Entry::Occupied(_) => {
                            Diagnostics::push(
                                db,
                                Diagnostic::new(
                                    src_decl.span().start,
                                    src_decl.span().end,
                                    format!("Duplicate type declaration {}", name.text(db)),
                                    module.filename.to_string_lossy().into(),
                                ),
                            );
                        }
                        Entry::Vacant(e) => {
                            e.insert(TypeDecl::Data(DataDecl {
                                type_: *type_,
                                name: abs_name,
                                params,
                                kind: kind.clone(),
                                constructors,
                            }));
                        }
                    }
                    iter.next();
                }

                TypeSynonym { name, params, body } => {
                    let abs_name = AbsoluteName::new(db, self.module_id, *name);

                    match self.types.entry(abs_name) {
                        Entry::Occupied(_) => {
                            Diagnostics::push(
                                db,
                                Diagnostic::new(
                                    src_decl.span().start,
                                    src_decl.span().end,
                                    format!("Duplicate type declaration {}", name.text(db)),
                                    module.filename.to_string_lossy().into(),
                                ),
                            );
                        }
                        Entry::Vacant(e) => {
                            let mut params = params.clone();
                            to_relative_params(abs_name, ref_loc, &mut params);

                            let mut body = body.clone();
                            body.to_relative_span(abs_name, ref_loc);

                            e.insert(TypeDecl::Type(TypeSynonymDecl {
                                name: abs_name,
                                params,
                                body,
                            }));
                        }
                    }
                    iter.next();
                }

                KindSignature { .. } => todo!(),

                Role(_) => todo!(),

                TypeSignature(sig) => {
                    iter.next();
                    match iter.peek() {
                        Some(x) if matches!(&****x, ValueDeclaration(_)) => {
                            self.parse_value_decl(
                                &mut iter,
                                Some(Located::new(src_decl.span(), sig.clone())),
                                ref_loc,
                            );
                        }
                        _ => {
                            Diagnostics::push(
                                db,
                                Diagnostic::new(
                                    src_decl.span().start,
                                    src_decl.span().end,
                                    format!(
                                        "Type signature of {} should be followed by its definition",
                                        sig.ident.text(db)
                                    ),
                                    module.filename.to_string_lossy().into(),
                                ),
                            );
                        }
                    }
                }
                ValueDeclaration(_) => {
                    self.parse_value_decl(&mut iter, None, ref_loc);
                }
                Destructuring { .. } => {
                    Diagnostics::push(
                        db,
                        Diagnostic::new(
                            src_decl.span().start,
                            src_decl.span().end,
                            "Invalid top-level destructuring".to_string(),
                            module.filename.to_string_lossy().into(),
                        ),
                    );
                    iter.next();
                }
                ForeignValue { name, type_ } => {
                    let abs_name = AbsoluteName::new(db, self.module_id, *name);

                    match self.values.entry(abs_name) {
                        Entry::Occupied(_) => {
                            Diagnostics::push(
                                db,
                                Diagnostic::new(
                                    src_decl.span().start,
                                    src_decl.span().end,
                                    format!("Duplicate value declaration {}", name.text(db)),
                                    self.filename.to_string_lossy().into(),
                                ),
                            );
                        }
                        Entry::Vacant(e) => {
                            let mut type_ = type_.clone();
                            type_.to_relative_span(abs_name, ref_loc);

                            e.insert(ValueDecl {
                                name: abs_name,
                                type_: Some(type_),
                                equations: vec![],
                            });
                        }
                    }
                    iter.next();
                }
                Class(type_class_decl) => {
                    let abs_name = AbsoluteName::new(db, self.module_id, type_class_decl.name);

                    match self.classes.entry(abs_name) {
                        Entry::Occupied(_) => {
                            Diagnostics::push(
                                db,
                                Diagnostic::new(
                                    src_decl.span().start,
                                    src_decl.span().end,
                                    format!(
                                        "Duplicate typeclass declaration {}",
                                        type_class_decl.name.text(db)
                                    ),
                                    self.filename.to_string_lossy().into(),
                                ),
                            );
                        }
                        Entry::Vacant(e) => {
                            let mut constraints = type_class_decl.constraints.clone();
                            constraints
                                .iter_mut()
                                .for_each(|c| c.to_relative_span(abs_name, ref_loc));

                            let mut params = type_class_decl.params.clone();
                            to_relative_params(abs_name, ref_loc, &mut params);

                            let mut methods = type_class_decl.methods.clone();
                            methods
                                .iter_mut()
                                .for_each(|decl| decl.r#type.to_relative_span(abs_name, ref_loc));

                            e.insert(TypeClassDecl {
                                name: abs_name,
                                constraints,
                                params,
                                fundeps: type_class_decl.fundeps.clone(),
                                methods,
                            });
                        }
                    }

                    iter.next();
                }
                InstanceChain(_) => todo!(),
                Operator { .. } => todo!(),
            }
        }
    }

    fn parse_value_decl<'m>(
        &mut self,
        iter: &mut Peekable<impl Iterator<Item = &'m Declaration>>,
        sig: Option<Located<TypeDeclarationData>>,
        reference_loc: usize,
    ) {
        let db = self.db;
        let Some(first_decl) = iter.next() else {
            panic!("parse_value_decl should be called when next is ValueDeclaration")
        };
        let DeclarationKind::ValueDeclaration(first) = &***first_decl else {
            panic!("parse_value_decl should be called when next is ValueDeclaration")
        };

        let abs_name = AbsoluteName::new(db, self.module_id, first.ident);

        let ty = match sig {
            None => None,
            Some(sig) => {
                if sig.ident == first.ident {
                    let mut sig_ty = sig.into_inner().r#type;
                    sig_ty.to_relative_span(abs_name, reference_loc);
                    Some(sig_ty)
                } else {
                    Diagnostics::push(
                        db,
                        Diagnostic::new(
                            sig.span().start,
                            sig.span().end,
                            format!(
                                "Type signature of {} should be followed by its definition",
                                sig.ident.text(db)
                            ),
                            self.filename.to_string_lossy().into(),
                        ),
                    );
                    None
                }
            }
        };
        let mut equations = vec![first.clone().into()];

        //TODO: Make spans relative in equations SourceSpan
        while let Some(src_decl) = iter.peek() {
            let DeclarationKind::ValueDeclaration(decl) = &****src_decl else {
                break;
            };
            if decl.ident != first.ident {
                // Start of another chain of equations
                break;
            }
            equations.push(decl.clone().into());
            iter.next();
        }

        match self.values.entry(abs_name) {
            Entry::Occupied(_) => {
                Diagnostics::push(
                    db,
                    Diagnostic::new(
                        first_decl.span().start,
                        first_decl.span().end,
                        format!("Duplicate value declaration {}", first.ident.text(db)),
                        self.filename.to_string_lossy().into(),
                    ),
                );
            }
            Entry::Vacant(e) => {
                e.insert(ValueDecl {
                    name: abs_name,
                    type_: ty,
                    equations,
                });
            }
        }
    }
}

fn to_relative_params(
    abs_name: AbsoluteName,
    reference_loc: usize,
    params: &mut [(crate::symbol::Symbol, Option<Type>)],
) {
    params.iter_mut().for_each(|(_, type_opt)| {
        if let Some(typ) = type_opt {
            typ.to_relative_span(abs_name, reference_loc)
        }
    })
}

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb)]
pub struct IndexedModule {
    pub module_id: ModuleId,
    pub types: FxHashMap<AbsoluteName, TypeDecl>,
    pub values: FxHashMap<AbsoluteName, ValueDecl>,
    pub classes: FxHashMap<AbsoluteName, TypeClassDecl>,
}

#[salsa::tracked]
pub fn indexed_module(db: &dyn Db, module_id: ModuleId) -> IndexedModule {
    let module = crate::parsed_module(db, module_id);
    let mut indexer = ModuleIndexer {
        db,
        types: Default::default(),
        values: Default::default(),
        classes: Default::default(),
        module_id,
        filename: module.filename.clone(),
    };
    indexer.index(module);
    IndexedModule {
        module_id,
        types: indexer.types,
        values: indexer.values,
        classes: indexer.classes,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::utils::tests::*;
    use indoc::indoc;
    use insta::{self, assert_snapshot};

    fn index_module(input: &str) -> String {
        let db = &mut crate::Database::test_single_file_db(input);
        let module_id = parse_module_id(input, db);
        format!(
            "{:#?}",
            (
                indexed_module(db, module_id).into_debug_all(db),
                indexed_module::accumulated::<Diagnostics>(db, module_id)
            )
        )
        .drop_salsa_id()
    }

    #[test]
    fn index_data_decl() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        data Foo = Foo | Bar
        "
        )));
    }

    #[test]
    fn duplicate_data_decl() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        data Foo
        data Foo
        "
        )));
    }

    #[test]
    fn type_sig_without_equations_1() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo :: Int
        "
        )));
    }

    #[test]
    fn type_sig_without_equations_2() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo :: Int
        bar = 1
        "
        )));
    }

    #[test]
    fn type_sig_without_equations_3() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo :: Int
        data Foo
        "
        )));
    }

    #[test]
    fn value_decl() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo = 1
        "
        )));
    }

    #[test]
    fn value_decl_with_signature() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo :: Int
        foo = 1
        "
        )));
    }

    #[test]
    fn value_decl_many_equations() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo 1 = 1
        foo 2 = 2
        "
        )));
    }

    #[test]
    fn type_alias() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        type Foo = Int
        "
        )));
    }

    #[test]
    fn type_and_data_decl_conflict() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        type Foo = Int
        data Foo
        "
        )));
    }

    #[test]
    fn foreign_import() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foreign import foo :: Int
        "
        )));
    }

    #[test]
    fn duplicate_value() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo = 1
        bar = 2
        foo = 1
        "
        )));
    }

    #[test]
    fn typeclass() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        
        class Show a where
          show :: a -> String
        "
        )));
    }
}
