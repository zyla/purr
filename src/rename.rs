use derive_new::new;

use crate::indexed_module::ValueDecl;
use crate::symbol::Symbol;
use crate::Db;
use std::collections::{HashMap, HashSet};

use crate::ast::*;
use crate::indexed_module::IndexedModule;
use crate::renamed_module::DeclId;
use crate::ModuleId;

pub fn rename_module(
    db: &dyn Db,
    module: &mut IndexedModule,
    imported_decls: Vec<(Option<ModuleId>, DeclId)>,
    exported_decls: Vec<DeclId>,
    diagnostics: &mut Vec<Diagnostic>,
) {
    let exported = exported_decls.iter().map(|decl_id| {
        (
            QualifiedName::new(db, Option::None, decl_id.name),
            AbsoluteName::new(db, decl_id.module, decl_id.name),
        )
    });

    let module_scope = imported_decls
        .iter()
        .map(|(qualified_as, id)| {
            (
                QualifiedName::new(db, *qualified_as, id.name),
                AbsoluteName::new(db, id.module, id.name),
            )
        })
        .chain(exported)
        .collect::<HashMap<_, _>>();
    let mut r = Renamer {
        db,
        module_scope,
        local_scopes: vec![HashSet::new()],
        diagnostics: vec![],
    };
    module.rename(&mut r);

    *diagnostics = r.diagnostics;
}

#[derive(Clone, Debug, new)]
pub struct Diagnostic {
    pub name: Symbol,
    pub error: String,
}

struct Renamer<'db> {
    db: &'db dyn Db,
    /// Maps from names as appear in source code to actual absolute names
    module_scope: HashMap<QualifiedName, AbsoluteName>,
    local_scopes: Vec<HashSet<Symbol>>,
    diagnostics: Vec<Diagnostic>,
}

impl<'db> Renamer<'db> {
    fn push_scope(&mut self) {
        self.local_scopes.push(HashSet::new());
    }

    fn pop_scope(&mut self) {
        let scope = self.local_scopes.pop();
        assert!(scope.is_some(), "pop_scope called when there are no scopes");
    }

    fn top_scope(&mut self) -> &mut HashSet<Symbol> {
        self.local_scopes
            .last_mut()
            .expect("top_scope called when there are no scopes")
    }

    fn push_diagnostic(&mut self, diagnostic: Diagnostic) {
        self.diagnostics.push(diagnostic)
    }
}

trait Rename {
    fn rename(&mut self, r: &mut Renamer);
}

impl<T> Rename for Option<T>
where
    T: Rename,
{
    fn rename(&mut self, r: &mut Renamer) {
        match self {
            None => {}
            Some(x) => x.rename(r),
        }
    }
}

impl<T> Rename for Located<T>
where
    T: Rename,
{
    fn rename(&mut self, r: &mut Renamer) {
        self.1.rename(r);
    }
}

impl Rename for IndexedModule {
    fn rename(&mut self, r: &mut Renamer) {
        self.values.iter_mut().for_each(|(_, ref mut v)| {
            v.rename(r);
        });

        // TODO: self.types
        // TODO: self.classes
        assert!(self.types.is_empty(), "renaming types not yet supported")
    }
}

impl Rename for ValueDecl {
    fn rename(&mut self, r: &mut Renamer) {
        self.type_.rename(r);
        self.equations.iter_mut().for_each(|ref mut x| x.rename(r));
    }
}

impl Rename for Type {
    fn rename(&mut self, _r: &mut Renamer) {
        // TODO
    }
}

impl Rename for CaseBranch {
    fn rename(&mut self, r: &mut Renamer) {
        r.push_scope();
        for ref mut pat in &mut self.pats {
            pat.rename(r);
        }
        self.expr.rename(r);
        r.pop_scope();
    }
}

impl Rename for PossiblyGuardedExpr {
    fn rename(&mut self, r: &mut Renamer) {
        match self {
            Self::Unconditional(ref mut e) => e.rename(r),
            Self::Guarded(_) => todo!("Pattern guards not implemented"),
        }
    }
}

impl Rename for PatKind {
    fn rename(&mut self, r: &mut Renamer) {
        match self {
            Self::Var(v) => {
                if !r.top_scope().insert(*v) {
                    r.push_diagnostic(
                        Diagnostic::new(*v, format!("Duplicate viariable '{}' in pattern", v.text(r.db)))
                    );
                }
            }
            _ => todo!("renaming PatKind {:?} not supported", self),
        }
    }
}

impl Rename for ExprKind {
    fn rename(&mut self, r: &mut Renamer) {
        match self {
            Self::Var(ref mut v) => {
                let db = r.db;
                let local_vars = r.top_scope();
                let is_local = v.module(db).is_none() && local_vars.contains(&v.name(db));
                if !is_local {
                    match r.module_scope.get(&v) {
                        None => {
                            r.push_diagnostic(
                                Diagnostic::new(v.name(db), format!("Unknown viariable '{}'", v.name(db).text(db)))
                            );
                        }

                        Some(abs) => {
                            *v = abs.to_qualified_name(db);
                            use salsa::DebugWithDb;
                            eprintln!("replacing with {:?}", v.clone().into_debug_all(db));
                        }
                    }
                }
            }
            Self::Lam(ref mut pats, ref mut expr) => {
                r.push_scope();
                for ref mut pat in pats {
                    pat.rename(r);
                }
                expr.rename(r);
                r.pop_scope();
            }
            Self::App(ref mut expr, ref mut exprs) => {
                expr.rename(r);
                for ref mut expr in exprs {
                    expr.rename(r);
                }
            }
            _ => todo!("renaming ExprKind {:?} not supported", self),
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use indoc::indoc;
    use insta::{self, assert_snapshot};
    use salsa::DebugWithDb;

    fn rename_mod(input: &str) -> String {
        let db = &mut crate::Database::test_single_file_db(input);
        let module_id = ModuleId::new(db, "Test".into());

        let lib = indoc!(
            "
        module Lib where
        
        data A = A
        data B = B

        a :: A
        a = A

        b :: B
        b = B

        "
        );
        db.add_source_file("lib.purs".into(), lib.into()).unwrap();

        let lib2 = indoc!(
            "
        module Lib2 where
                
        import Lib        

        f :: A -> B -> A
        f a _ = a
        "
        );
        db.add_source_file("Lib2.purs".into(), lib2.into()).unwrap();

        let mut module = crate::indexed_module::indexed_module(db, module_id);
        let imported = crate::renamed_module::imported_decls(db, module_id);
        let exported = crate::renamed_module::exported_decls(db, module_id);
        let mut diagnostics = vec![];

        rename_module(db, &mut module, imported, exported, &mut diagnostics);

        format!("{:#?}", (module.into_debug_all(db), diagnostics))
    }

    #[test]
    fn smoke() {
        assert_snapshot!(rename_mod(indoc!(
            "
        module Test where
        
        f a = a 
        "
        )))
    }

    #[test]
    fn some_modules() {
        assert_snapshot!(rename_mod(indoc!(
            "
        module Test where 
        
        import Lib
        import Lib2 

        g :: A -> B -> A
        g a b = f a b

        h :: A
        h = g a b
        "
        )))
    }

    #[test]
    fn duplicate_var_in_pattern() {
        assert_snapshot!(rename_mod(indoc!(
            "
        module Test where
        
        f a a = a
        "
        )))
    }

    #[test]
    fn unknown_var() {
        assert_snapshot!(rename_mod(indoc!(
            "
        module Test where
        
        g = x
        "
        )))
    }
}
