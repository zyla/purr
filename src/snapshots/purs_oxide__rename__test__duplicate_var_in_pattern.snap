---
source: src/rename.rs
expression: "rename_mod(indoc!(\"\n        module Test where\n        \n        f a a = a\n        \"))"
---
(
    IndexedModule {
        module_id: ModuleId {
            [salsa id]: 0,
            name: "Test",
        },
        types: {},
        values: {
            AbsoluteName {
                [salsa id]: 0,
                module: ModuleId {
                    [salsa id]: 0,
                    name: "Test",
                },
                name: Symbol {
                    [salsa id]: 0,
                    text: "f",
                },
            }: ValueDecl {
                name: AbsoluteName {
                    [salsa id]: 0,
                    module: ModuleId {
                        [salsa id]: 0,
                        name: "Test",
                    },
                    name: Symbol {
                        [salsa id]: 0,
                        text: "f",
                    },
                },
                type_: None,
                equations: [
                    CaseBranch {
                        pats: [
                            Located(
                                SourceSpan {
                                    start: 21,
                                    end: 22,
                                },
                                Var(
                                    Symbol {
                                        [salsa id]: 1,
                                        text: "a",
                                    },
                                ),
                            ),
                            Located(
                                SourceSpan {
                                    start: 23,
                                    end: 24,
                                },
                                Var(
                                    Symbol {
                                        [salsa id]: 1,
                                        text: "a",
                                    },
                                ),
                            ),
                        ],
                        expr: Unconditional(
                            Located(
                                SourceSpan {
                                    start: 27,
                                    end: 28,
                                },
                                Var(
                                    QualifiedName {
                                        [salsa id]: 0,
                                        module: None,
                                        name: Symbol {
                                            [salsa id]: 1,
                                            text: "a",
                                        },
                                    },
                                ),
                            ),
                        ),
                    },
                ],
            },
        },
        classes: {},
    },
    [
        Diagnostic {
            name: Symbol(
                Id {
                    value: 2,
                },
            ),
            error: "Duplicate variable 'a' in pattern",
        },
    ],
)
